didel/courses.py

```diff
@@ -146,3 +157,76 @@ def unenroll(self):
         text = u'Vous avez été désinscrit'
         params = {'cmd': 'exUnreg', 'course': self.ref}
         return self.session.get_ensure_text(path, text, params=params)
+
+
+class DocumentsLinks(DidelEntity):
+
+    URL_FMT = '/claroline/document/document.php?cidReset=true&cidReq={ref}'
+
+    def __init__(self, ref, sub_url=None):
+        self.ressources = {}
+        self.ref = ref
+        if(sub_url is None):
+            self.path = self.URL_FMT.format(ref=ref)
+        else:
+            self.path = sub_url
+            self.ref = ""
+        super(DocumentsLinks, self).__init__(self.ref)
+
+
+    def populate(self, soup, session):
+        """
+        get all documents and folder of a course
+        """
+        table = soup.select(".claroTable tbody tr[align=center]")
+        for line in table:
+            cols = line.select("td")
+            item = cols[0].select(".item")[0]
+            document_nom = item.contents[1].strip()
+            document_date = cols[2].select("small")[0].contents[0].strip()
+            document_href = cols[0].select("a")[0].attrs["href"].strip()
+            if(re.match(r"^<img (alt=\"\")? src=\"/web/img/folder", str(item.select("img")[0]))) is not None:
+                doc = DocumentsLinks("", document_href)
+                doc.fetch(self.session)
+                self.ressources[document_nom] = doc
+            else:
+                self.ressources[document_nom] = Document(document_nom, document_href, document_date)
+
+    def timestamp(self, date):
+        return mktime(datetime.datetime.strptime(date, "%d.%m.%Y").timetuple())
+ 
+    def synchronize(self, path):
+        """
+        compare files on didels with file in folder,
+            and (calling download add or reset files'user
+            only if not exist or older
+        """
+        path = path + "/" + self.ref
+        if not(os.path.isdir(path)):
+            mkdir(path)
+        for k in self.ressources:
+            if(isinstance(self.ressources[k], DocumentsLinks)): 
+                self.ressources[k].synchronize(path + "/" + k)
+            else:
+                if (not os.path.exists(path + "/" + k) or (
+                        (os.stat(path + "/" + k).st_mtime <
+                        self.timestamp(self.ressources[k].date))
+                    )):
+                    self.download(self.ressources[k], path)
+
+    def download(self, document, path):
+        response = self.session.get(ROOT_URL + document.href)
+        document.path = path + "/" + document.nom
+        file = open(document.path, 'w')
+        file.write(response.content)
+        file.close()
+        t = self.timestamp(document.date)
+        os.utime(document.path, (t, t))
+        print(document.path + "... downloaded")
+
+class Document:
+
+    def __init__(self, nom, href, date):
+        self.nom = nom
+        self.href = href
```

Comment:
What does `href` mean here? If it’s a path, name it `path`. If it’s a full URL, use `url`.